### ISSUE TITLE
The texture section  in file userguide.md should document calling the tobytes() method on numpy arrays used for it's data.

### ISSUE THREAD

**Issue by ron-adam** [datoviz/datoviz#71]
Textures expect a data stream that match the Vulkan format type specified. 

A typical numpy image array may have the following form:
A numpy array with a shape of (800, 600, 4).  That has a dtype of 'float32'.
Each of the (4) values are in the range of 0.0 to 0.1 for the color values r,g,b, and a.

```
    width, height = data.shape[:2]
   
    # A 32 bit colors range 0.0 to 1.0.
    format = dvz.FORMAT_R32G32B32A32_SFLOAT
    image =data

    # Or an 8 bit colors 0 to 255
    format = dvz.FORMAT_R8G8B8A8_UNORM
    image = np.around(data * 255).astype('uint8')

    tex = dvz.tex_image(batch, format, width, height, A_(image.tobytes()))    #<--- tobytes() needed here.
```
The numpy array tobytes() method results in a byte stream that matches the formats in both of these cases.  The width and height on the texture determine the how the stream is mapped to the texture.   The A_() wrapper may not be needed in this case.

It took me quite a while to remember I had run across this before.   I suggest showing it's use in both the image.py example, and the volume example.   Also adding it to the texture example in userguide.md.   That should save others quiet a lot of searching after their textures don't work like expected.  This is not explained well elsewhere, or at least not in an easy to find place.

**Comment by rossant** [comment]
Do you mean you're unable to use textures without calling `.tobytes()`? Aren't the image and volume Python examples working for you?

In principle, the builtin `A_()` wrapper should take care of the conversion from NumPy to a pointer as expected by the wrapped C functions, without needing `tobytes()`, but I may have missed something.

In any case, the texture-related documentation in `userguide.md` should clearly be improved, I'll do that.

**Comment by ron-adam** [comment]
The problem appears when I attempt to use my own generated data in numpy arrays.  The lines above worked for me, while just passing the array as A_(image) did not.   It shows an incorrect image with elements that repeat and are of the wrong color and scale. 

Just tested agian.  using image.revel() also works correctly. 
And checking... np.array(image, order='C') works too, but not just sending the array directly.

Ok... I see what's going on.  I'm generating the data with cupy and numba, which apparently use fortran array row, column ordering.
The ravel() and tobyte() methods probably default to 'C' order when they flatten the data.

So... possibly only needs a small note in the docs to check to that the array order='C'?


A bit off topic....
My use case is to generate images and video on the GPU with Cupy and Numba, then send it to the display and video file object via queues.   This allows the simulation, video and file objects to not need to wait on each other.   I use a Vispy window currently and ImageIO.  

The issue I have now is datoviz doesn't yet have a texture update call exposed to python.   (That I know of)  And it would be really nice if it could handle Numba/Cupy arrays.   I also wanted to look into what raytracing capabilities datoviz may have.   I currently write my own.

You can see an example of one of my volumetric wave simulations here:
      https://youtu.be/Jswky42HI74?si=5JKjRDfG40-7EYS3

In it, I'm using my own interpolation function to get smooth animation,  It may be of interest to you.

**Comment by rossant** [comment]
> Ok... I see what's going on. I'm generating the data with cupy and numba, which apparently use fortran array row, column ordering. The ravel() and tobyte() methods probably default to 'C' order when they flatten the data.
> 
> So... possibly only needs a small note in the docs to check to that the array order='C'?

Oh, you're totally right. Yes, the library definitely expects C ordering. Let me add this to the documentation.

> The issue I have now is datoviz doesn't yet have a texture update call exposed to python. (That I know of)

There is, but it appears to be still undocumented... A call like the following should work:

```
dvz.upload_tex(batch, tex, uvec3(0), uvec3(0), width*height*4, A_(image), 0)
# arguments: batch, texture ID, offset, shape (0=the entire texture), size in bytes of the next argument, pointer to the buffer to upload to the texture
```

> And it would be really nice if it could handle Numba/Cupy arrays.

Great idea! I'll look into it. It looks like CuPy supports "[UnownedMemory](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.UnownedMemory.html)" where CuPy arrays are created from a CUDA pointer, and there might be a way in Vulkan to create a memory block that is exposed to CUDA, allowing one to retrieve a CUDA pointer (then passed to CuPy). I'll have to experiment with this. It would be really cool to have a way to manipulate a GPU texture used in Datoviz with CuPy/Numba.

> I also wanted to look into what raytracing capabilities datoviz may have. I currently write my own.

There is a basic volume rendering example that uses raymarching. I have also experimented with GLSL-based GPU raytracing demos in the past.

Your video is really cool! I'd be glad to help you support this kind of use-case with Datoviz. Perhaps you would be interesting in writing a custom visual with custom shaders. There is a [very basic example here](https://github.com/datoviz/datoviz/blob/dev/examples/drp.py).

**Comment by ron-adam** [comment]
Hmmm... looks like I may need to install the dev branch version to get those to work.  I did a pip install datoviz, so don't have the dvz.upload_tex function.   The drp.py file gives me an error on the dvz.create_canvas function.  Also the mouse brain volume data is missing. 

I've been attempting to use the stent.npz file available from imageio.  This has a shape that isn't a cube and creates artifacts on the sides and back when it's scaled to correctly.  As long as I use dvz.volume_size(visual, 1,1,1)  I get a cube with no artifacts, but not the correct scale for the volume.

![image](https://github.com/user-attachments/assets/9dafd5d7-74f0-4bbb-844f-d2ab7a45b7ff)

And would you prefer I open separate issues for each question or possible bug.   Or maybe there is a general discussion area where we can discuss things more generally and open issues only if needed?   It looks like it's still pretty early in development, so maybe making progress is the priority however it happens?

Apparently it's very difficult to share OpenGL memory, so using Cupy with OpenGL directly hasn't been done.   Maybe it will be easier with the Vulkan api, and it may help that you are using C for the lower level components.

I'm still poking around in dataviz to get a feel for how it's organized and what it can do.

**Comment by rossant** [comment]
> Hmmm... looks like I may need to install the dev branch version to get those to work. I did a pip install datoviz, so don't have the dvz.upload_tex function. The drp.py file gives me an error on the dvz.create_canvas function. Also the mouse brain volume data is missing.

Yes, sorry, you need the dev branch and build it from source. I should do a new release soon.

> I've been attempting to use the stent.npz file available from imageio. This has a shape that isn't a cube and creates artifacts on the sides and back when it's scaled to correctly. As long as I use dvz.volume_size(visual, 1,1,1) I get a cube with no artifacts, but not the correct scale for the volume.

Yes, please feel free to open a new issue.

> And would you prefer I open separate issues for each question or possible bug. Or maybe there is a general discussion area where we can discuss things more generally and open issues only if needed?

Yes, maybe the Discussions are is good for general questions.

> It looks like it's still pretty early in development, so maybe making progress is the priority however it happens?

Yes, it is! the core Vulkan technology took a while to make right (I started experimenting with Vulkan 5 years ago, and the first release was in 2021), but I think the current architecture is robust enough now. I'm starting to see general discussions around about what the successor of Vulkan may look like. My main objective when designing Datoviz was to make sure the lowest-level bits are as independent as possible from the rest, so that moving to different/newer rendering technologies should not be totally impracticable.

I am actively working on Datoviz and I have funding for at least a few years. There are few users right now, which allows me to move fast and not worry too much about backward compatibility. My priority is to ensure stability and to implement important features for early users. Datoviz is going to be main rendering backend for the upcoming VisPy 2.0 library that I'm developing with @rougier.

I'm looking forward for your comments and suggestions.

**Comment by ron-adam** [comment]
Just an additional comment about array data order.

Your textures want specifically data in "C_CONTIGUOUS' order.
The order can change if the axis is moved to reshape them, and possibly by other opperations.
Larger arrays may not be contiguous.

It may be good to add something like this in places where it can matter. 
```python
    if not array.flags['C_CONTIGUOUS']:
        # If not, make a copy in C-order
        array = np.ascontiguousarray(array)
```

>> "I'm looking forward for your comments and suggestions."

I'm looking forward to learning more about Datoviz and Vispy2.0.  👍

I know some C code, but it's been awhile.  (And a bit of several other languages.)  I was involved in the development process for python3.0, Mostly in discussion on the dev and ideas lists.   So I know python very well.   As for graphics advice, I can share my own experience in rendering 3D volumetric data, and how I use that to create VR360 video.

**Comment by rossant** [comment]
> Your textures want specifically data in "C_CONTIGUOUS' order. The order can change if the axis is moved to reshape them, and possibly by other opperations. Larger arrays may not be contiguous.
> 
> It may be good to add something like this in places where it can matter.

Absolutely. The ctypes bindings actually already have this requirement for most functions, for example:

```python
marker_position.argtypes = [
    ctypes.POINTER(DvzVisual),  # DvzVisual* visual
    ctypes.c_uint32,  # uint32_t first
    ctypes.c_uint32,  # uint32_t count
    ndpointer(dtype=np.float32, ndim=2, ncol=3, flags="C_CONTIGUOUS"),  # vec3* values
    ctypes.c_int,  # int flags
]
```

However, I now realize that some of the functions do not have this requirement as they just expect a generic `ctypes.c_void_p`, like for example:

```python
tex_volume.argtypes = [
    ctypes.POINTER(DvzBatch),  # DvzBatch* batch
    DvzFormat,  # DvzFormat format
    ctypes.c_uint32,  # uint32_t width
    ctypes.c_uint32,  # uint32_t height
    ctypes.c_uint32,  # uint32_t depth
    ctypes.c_void_p,  # void* data
]
```

I think it would be cleaner to replace these by an `ndpointer(flags="C_CONTIGUOUS")`.

**Comment by rossant** [comment]
Just pushed a potential fix too here, on dev.

**Comment by ron-adam** [comment]
Looks like this is fixed, so closing.   (I presume that I can close issues I've opened, let me know if I shouldn't.)

**Comment by rossant** [comment]
great, thanks! (of course, feel free to close issues!)
